reactor-core/src/main/java/reactor/core/publisher/FluxIndexedFuseable.java

```diff
@@ -0,0 +1,385 @@
+		if (actual instanceof ConditionalSubscriber) {
+			@SuppressWarnings("unchecked") ConditionalSubscriber<? super I> cs =
+					(ConditionalSubscriber<? super I>) actual;
+			source.subscribe(new IndexedFuseableConditionalSubscriber<>(cs, indexMapper));
+		}
+		else {
+			source.subscribe(new IndexedFuseableSubscriber<>(actual, indexMapper));
+		}
+	}
+
+	static final class IndexedFuseableSubscriber<I, T> implements InnerOperator<T, I>,
+	                                                              QueueSubscription<I> {
+
+		final CoreSubscriber<? super I>             actual;
+		final BiFunction<? super Long, ? super T, ? extends I> indexMapper;
+
+		boolean              done;
+		long                 index;
+		QueueSubscription<T> s;
+		int                  sourceMode;
+
+		IndexedFuseableSubscriber(CoreSubscriber<? super I> actual,
+				BiFunction<? super Long, ? super T, ? extends I> indexMapper) {
+			this.actual = actual;
+			this.indexMapper = indexMapper;
+		}
+
+		@Override
+		public void onSubscribe(Subscription s) {
+			if (Operators.validate(this.s, s)) {
+				//noinspection unchecked
+				this.s = (QueueSubscription<T>) s;
+
+				actual.onSubscribe(this);
+			}
+		}
+
+		@Override
+		@Nullable
+		public I poll() {
+			T v = s.poll();
+			if (v != null) {
+				long i = this.index;
+				I indexed = Objects.requireNonNull(indexMapper.apply(i, v),
+						"indexMapper returned a null value at raw index " +
+								i + " for value " + v);
+				this.index = i + 1;
+				return indexed;
+			}
+			return null;
+		}
+
+		@Override
+		public void onNext(T t) {
+			if (sourceMode == ASYNC) {
+				actual.onNext(null);
+			}
+			else {
+				if (done) {
+					Operators.onNextDropped(t, actual.currentContext());
+					return;
+				}
+
+				long i = this.index;
+				try {
+					I indexed = Objects.requireNonNull(indexMapper.apply(i, t),
+							"indexMapper returned a null value at raw index " + i +
+									" for value " + t);
+					this.index = i + 1L;
+					actual.onNext(indexed);
+				}
+				catch (Throwable e) {
+					onError(Operators.onOperatorError(s, e, t, actual.currentContext()));
+				}
+			}
+		}
+
+		@Override
+		public void onError(Throwable t) {
+			if (done) {
+				Operators.onErrorDropped(t, actual.currentContext());
+				return;
+			}
+
+			done = true;
+
+			actual.onError(t);
+		}
+
+		@Override
+		public void onComplete() {
+			if (done) {
+				return;
+			}
+			done = true;
+
+			actual.onComplete();
+		}
+
+		@Override
+		public CoreSubscriber<? super I> actual() {
+			return this.actual;
+		}
+
+		@Override
+		public void request(long n) {
+			s.request(n);
+		}
+
+		@Override
+		public void cancel() {
+			s.cancel();
+		}
+
+		@Override
+		public boolean isEmpty() {
+			return s.isEmpty();
+		}
+
+		@Override
+		public void clear() {
+			s.clear();
+		}
+
+		@Override
+		public int requestFusion(int requestedMode) {
+			int m;
+			if ((requestedMode & Fuseable.THREAD_BARRIER) != 0) {
```

Comment:
if this was using Tuple.of it could avoid forcing Thread Barrier condition (not user code). Its a minor optimization but one more reason to have a static for Tuples::of

Reply:
mmh the thing is the `BiFunction` determines the output type of the `Flux`, it is not necessarily a `Tuple2` (despite the class comment, which I should change).

@smaldini Do you mean to revert back to when the function was just about producing a user-facing index, aka the left part of the `Tuple2`, and the operator would statically map to `Tuple2`?